xla/service/gpu/cudnn_norm_rewriter.cc

```diff
@@ -0,0 +1,629 @@
+template <typename Pattern0, typename Pattern1>
+auto MultiplyAnyOrder(HloInstruction** multiply, Pattern0 pattern0,
+                      Pattern1 pattern1) {
+  return OptionalConvertAndOrReshape(
+      m::MultiplyAnyOrder(multiply, pattern0, pattern1));
+}
+
+// Multiplication of pattern by itself with optional convert and/or reshape.
+template <typename Pattern>
+auto Square(Pattern pattern) {
+  return MultiplyAnyOrder(pattern, pattern)
+      .WithPredicate([](const HloInstruction* instr) {
+        return instr->unique_operands().size() == 1;
+      });
+}
+
+// Reduction-addition of pattern with optional convert and/or reshape and
+// constant 0 scalar.
+template <typename Pattern>
+auto ReduceAdd(Pattern pattern) {
+  return OptionalConvertAndOrReshape(
+      m::Reduce(pattern, m::ConstantScalar(0))
+          .WithPredicate(
+              [](const HloInstruction* instr) { return IsReduceAdd(instr); }));
+}
+
+// Capturing reduction-addition of pattern with optional convert and/or reshape
+// and constant 0 scalar.
+template <typename Pattern>
+auto ReduceAdd(HloInstruction** reduction, Pattern pattern) {
+  return OptionalConvertAndOrReshape(
+      m::Reduce(reduction, pattern, m::ConstantScalar(0))
+          .WithPredicate(
+              [](const HloInstruction* instr) { return IsReduceAdd(instr); }));
+}
+
+// Expected value, or mean, with optional broadcast.
+template <typename Pattern>
+auto Expectation(Pattern pattern) {
+  auto shared_subpattern =
+      MultiplyAnyOrder(m::Broadcast(m::ConstantScalar()), ReduceAdd(pattern));
+  return m::AnyOf<HloInstruction>(m::Broadcast(shared_subpattern),
+                                  shared_subpattern);
+}
+
+// Expected value, or mean, with optional broadcast.
+template <typename Pattern>
+auto Expectation(HloInstruction** expectation, Pattern pattern) {
+  auto shared_subpattern = MultiplyAnyOrder(
+      expectation, m::Broadcast(m::ConstantScalar()), ReduceAdd(pattern));
+  return m::AnyOf<HloInstruction>(m::Broadcast(shared_subpattern),
+                                  shared_subpattern);
+}
+
+// Expected value, or mean, with optional broadcast.
+template <typename Pattern>
+auto Expectation(HloInstruction** expectation, HloInstruction** reduction,
+                 Pattern pattern) {
+  auto shared_subpattern =
+      MultiplyAnyOrder(expectation, m::Broadcast(m::ConstantScalar()),
+                       ReduceAdd(reduction, pattern));
+  return m::AnyOf<HloInstruction>(m::Broadcast(shared_subpattern),
+                                  shared_subpattern);
+}
+
+// Variance, expressed as E[(X - E[X])^2] or E[X^2] - E[X]^2 where E is the
+// expecation.
+auto Variance(HloInstruction** input0, HloInstruction** input1) {
+  return m::AnyOf<HloInstruction>(
+      Subtract(Expectation(Square(m::Op(input0))),
+               Square(Expectation(m::Op(input1)))),
+      Expectation(Square(Subtract(m::Op(input0), Expectation(m::Op(input1))))));
+}
```

Comment:
@reedwm

TLDR: I think that currently the only good option is to compare the HLO pointers after the matching.

More thoughts (Sorry, this was interesting to me so I wrote a lot):

Let's say you want to match (A - A). The following kindof works:

(**Not recommended code:**)
```
auto SameOp(HloInstruction** existing_match) {
  return m::Op().WithPredicate(
      [existing_match](const HloInstruction* new_match) {
        return new_match == *existing_match;
      });
}

m::Subtract(m::Op(&lhs), SameOp(&lhs))
```

But this depends on the order of matching. For example if the Subtract pattern would match the operands in reverse order, it would stop working. Also, this reordering could very well happen with AnyOf, AnyOrder, etc. So I would not recommend doing this.

I think that to get a better solution we would have to implement a more complex matching algorithm which can detect whether a pattern is being matched the first time and behave differently based on that.
The tricky part is handling AnyOf, because if one of the branches fail, we have to forget all the matches that we found in that branch.

But if the Match function returned a map of matches instead of setting pointers, this could actually work.
The Match function's declaration could look like this:
```
std::pair<bool, PatternToHloInstructionMap> Match(PatternToHloInstructionMap matches_so_far);
```

I implemented this for NNEF Tools a few years ago, but in python:
https://github.com/KhronosGroup/NNEF-Tools/blob/5c928df05994ac3894512b5428513709bc72ad6d/nnef_tools/core/matcher.py#L306
Tensors can be matched multiple times with this.
The usage was something like this:
https://github.com/KhronosGroup/NNEF-Tools/blob/5c928df05994ac3894512b5428513709bc72ad6d/nnef_tools/optimization/data_format_optimizer.py#L321C1-L322C1

If you think that a new matcher library would be useful, feel free to create a bug, but whether we can do it depends on prioritization.

Reply:
I think that it's not good that the `HloInstruction* lhs, rhs;` are local to the SubtractFromSelf function, because the patterns will be matched after that function returns, so those variables will not be alive. Also the "instr" is unused.

> Would the WithPredicate be evaluated after lhs and rhs are assigned?
I think that it would. (WithPredicate defines an AllofPattern and that evaluates its parameters in order.)

(Honestly I would recommend just using the patterns in the existing way.)

Reply:
Hmm this might be possible by capturing an HloInstruction in the lambda. But for now, I'm fine with the PR's current approach.

@philipphack can you add a comment stating the caller should check that input0 is the same as input1?